### DISCUSSION TITLE
Is each field meant to be validated by at most one validator?

### DISCUSSION THREAD

**Discussion by DMRobertson** [pydantic/pydantic#4087] [Question] [answered]
I've just tried the following with Pydantic 1.9.0.

```python
from pydantic import *
class TestModel(BaseModel):
    x: StrictStr
    @validator("x")
    def validate_x_one(cls, value):
        print("validator 1")
        return ValueError("bad 1")
    @validator("x")
    def validate_x_two(cls, value):
        print("validator 2")
        raise ValueError("bad 2")

TestModel(x="")
```

Pasting this into an interpreter, I see:

```
>>> TestModel(x="")
validator 1
validator 2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 331, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for TestModel
x
  bad 2 (type=value_error)

```

I was surprised that `bad 1` didn't show up in the output! I think this suggests that the first validator ran and raised, but its exception was discarded. Is this intended/expected behaviour?

(For context, I was hoping to use Pydantic to validate a config file. I wanted to have multiple small, independent checks on a given field so that they each be reported to the user in one batch.)

**Comment by DMRobertson** [comment]
I made a typo: `validate_x_one` returns its `ValueError` rather than raising it. The example should have read:

```python
from pydantic import *
class TestModel(BaseModel):
    x: StrictStr
    @validator("x")
    def validate_x_one(cls, value):
        print("validator 1")
        raise ValueError("bad 1")
    @validator("x")
    def validate_x_two(cls, value):
        print("validator 2")
        raise ValueError("bad 2")

TestModel(x="")
```

This now yields:

```
>>> from pydantic import *
>>> class TestModel(BaseModel):
...     x: StrictStr
...     @validator("x")
...     def validate_x_one(cls, value):
...         print("validator 1")
...         raise ValueError("bad 1")
...     @validator("x")
...     def validate_x_two(cls, value):
...         print("validator 2")
...         raise ValueError("bad 2")
... 
>>> TestModel(x="")
validator 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 331, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for TestModel
x
  bad 1 (type=value_error)
```

which makes more sense. However, it seems like the first error is reported and the second validator is never run. Is there some way to guarantee that both validators run and have both of their errors reported?

**Comment by DMRobertson** [comment, marked as answer]
Ahh, judging by https://github.com/samuelcolvin/pydantic/blob/8846ec4685e749b93907081450f592060eeb99b1/pydantic/fields.py#L1116-L1124 it looks like we stop validating a field as soon as one of its validators reports an error.

**Reply by samuelcolvin** [reply]
Yes exactly

**Reply by DMRobertson** [reply]
Thanks. Would the project be interested in a PR which changes this behaviour? (I think I've read somewhere that pydantic _isn't_ trying to be a generic validation library, so not sure if this kind of behaviour is in scope.)

**Reply by samuelcolvin** [reply]
have a look at [pydantic-core](https://github.com/samuelcolvin/pydantic-core) which will become the validation engine for pydantic in v2.

There you can use "wrap validators" which are like the "middleware onion", e.g. they take the next validator as an argument and optionally call it, that means you can catch errors and extend them.

There's some more work to do on this to make it more flexible, but I would't accept fundamental changes to validator logic to mean they all ran before returning an error.

**Reply by DMRobertson** [reply]
Wrap validators sound ideal. Many thanks!
